exchanges/poloniex/poloniex_websocket.go

```diff
@@ -28,1128 +29,503 @@ import (
 	}
 	return nil
 }
 
-// WsProcessOrderbookSnapshot processes a new orderbook snapshot into a local
-// of orderbooks
-func (e *Exchange) WsProcessOrderbookSnapshot(data []any) error {
-	subDataMap, ok := data[1].(map[string]any)
-	if !ok {
-		return fmt.Errorf("%w subData element is not map[string]any",
-			errTypeAssertionFailure)
-	}
-
-	pMap, ok := subDataMap["currencyPair"]
-	if !ok {
-		return errors.New("could not find currency pair in map")
-	}
-
-	pair, ok := pMap.(string)
-	if !ok {
-		return fmt.Errorf("%w subData element is not map[string]any",
-			errTypeAssertionFailure)
-	}
-
-	if len(data) < 3 {
-		return fmt.Errorf("%w for pair %v", errNotEnoughData, pair)
-	}
-
-	ts, ok := data[2].(string)
-	if !ok {
-		return common.GetTypeAssertError("string", data[2], "timestamp string")
-	}
-
-	tsMilli, err := strconv.ParseInt(ts, 10, 64)
-	if err != nil {
+func (e *Exchange) wsHandleData(ctx context.Context, conn websocket.Connection, respRaw []byte) error {
+	var result SubscriptionResponse
+	if err := json.Unmarshal(respRaw, &result); err != nil {
 		return err
 	}
-
-	oMap, ok := subDataMap["orderBook"]
-	if !ok {
-		return errors.New("could not find orderbook data in map")
-	}
-
-	ob, ok := oMap.([]any)
-	if !ok {
-		return fmt.Errorf("%w orderbook data is not []any",
-			errTypeAssertionFailure)
-	}
-
-	if len(ob) != 2 {
-		return errNotEnoughData
-	}
-
-	askData, ok := ob[0].(map[string]any)
-	if !ok {
-		return fmt.Errorf("%w ask data is not map[string]any",
-			errTypeAssertionFailure)
-	}
-
-	bidData, ok := ob[1].(map[string]any)
-	if !ok {
-		return fmt.Errorf("%w bid data is not map[string]any",
-			errTypeAssertionFailure)
-	}
-
-	var book orderbook.Book
-	book.Asks = make(orderbook.Levels, 0, len(askData))
-	for price, volume := range askData {
-		var p float64
-		p, err = strconv.ParseFloat(price, 64)
-		if err != nil {
-			return err
+	if result.ID != "" {
+		if !conn.IncomingWithData(result.ID, respRaw) {
+			return fmt.Errorf("could not match trade response with ID: %s Event: %s ", result.ID, result.Event)
 		}
-		v, ok := volume.(string)
-		if !ok {
-			return fmt.Errorf("%w ask volume data not string",
-				errTypeAssertionFailure)
-		}
-		var a float64
-		a, err = strconv.ParseFloat(v, 64)
-		if err != nil {
-			return err
-		}
-		book.Asks = append(book.Asks, orderbook.Level{Price: p, Amount: a})
+		return nil
 	}
-
-	book.Bids = make(orderbook.Levels, 0, len(bidData))
-	for price, volume := range bidData {
-		var p float64
-		p, err = strconv.ParseFloat(price, 64)
-		if err != nil {
-			return err
-		}
-		v, ok := volume.(string)
-		if !ok {
-			return fmt.Errorf("%w bid volume data not string",
-				errTypeAssertionFailure)
-		}
-		var a float64
-		a, err = strconv.ParseFloat(v, 64)
-		if err != nil {
-			return err
+	if result.Event != "" {
+		switch result.Event {
+		case "pong", "subscribe":
+		case "error":
+			if result.Message == "user must be authenticated!" {
```

Comment:
🤷‍♂️ Clarification please: Are you sure that this message is correct?
It's not shown in the logs.